File: web/lib/volunteer-opportunities.ts
import { createServiceClient } from "@/lib/supabase/service";
import type { AnySupabase } from "@/lib/api-utils";

export type VolunteerOpportunity = {
  id: string;
  slug: string;
  title: string;
  summary: string | null;
  description: string | null;
  commitment_level: "drop_in" | "ongoing" | "lead_role";
  time_horizon: "one_day" | "multi_week" | "multi_month" | "ongoing" | null;
  onboarding_level: "none" | "light" | "screening_required" | "training_required" | null;
  schedule_summary: string | null;
  location_summary: string | null;
  skills_required: string[];
  language_support: string[];
  physical_demand: "low" | "medium" | "high" | null;
  min_age: number | null;
  family_friendly: boolean;
  group_friendly: boolean;
  remote_allowed: boolean;
  accessibility_notes: string | null;
  background_check_required: boolean;
  training_required: boolean;
  capacity_total: number | null;
  capacity_remaining: number | null;
  urgency_level: "normal" | "urgent";
  starts_on: string | null;
  ends_on: string | null;
  application_url: string;
  source_url: string | null;
  metadata: Record<string, unknown>;
  is_active: boolean;
  updated_at: string;
  source_id: number | null;
  portal_id: string | null;
  event_id: number | null;
  organization: {
    id: string;
    name: string;
    slug: string;
    org_type: string | null;
    website: string | null;
    logo_url: string | null;
    description: string | null;
    city: string | null;
    neighborhood: string | null;
  };
  source: {
    id: number;
    name: string;
    slug: string;
    url: string;
  } | null;
  event: {
    id: number;
    title: string;
    start_date: string;
    start_time: string | null;
    end_time: string | null;
  } | null;
};

export type VolunteerProfile = {
  user_id: string;
  causes: string[];
  skills: string[];
  availability_windows: Array<Record<string, unknown>>;
  travel_radius_km: number | null;
  mobility_constraints: string | null;
  languages: string[];
  commitment_preference: "drop_in" | "ongoing" | "lead_role" | "mixed" | null;
};

export type RankedVolunteerOpportunity = VolunteerOpportunity & {
  fit_score: number;
  fit_reasons: string[];
};

export type VolunteerOpportunitySummary = {
  total: number;
  by_commitment_level: Record<VolunteerOpportunity["commitment_level"], number>;
  by_cause: Array<{
    cause: string;
    count: number;
  }>;
};

type PortalRow = {
  id: string;
  slug: string;
  name: string;
};

export type VolunteerOpportunityFilters = {
  portalSlug: string;
  limit?: number;
  commitmentLevel?: VolunteerOpportunity["commitment_level"];
  organizationSlug?: string;
  timeHorizon?: NonNullable<VolunteerOpportunity["time_horizon"]>;
  onboardingLevel?: NonNullable<VolunteerOpportunity["onboarding_level"]>;
  remoteAllowed?: boolean;
  cause?: string;
  query?: string;
};

const VOLUNTEER_OPPORTUNITY_SELECT = `
  id,
  slug,
  title,
  summary,
  description,
  commitment_level,
  time_horizon,
  onboarding_level,
  schedule_summary,
  location_summary,
  skills_required,
  language_support,
  physical_demand,
  min_age,
  family_friendly,
  group_friendly,
  remote_allowed,
  accessibility_notes,
  background_check_required,
  training_required,
  capacity_total,
  capacity_remaining,
  urgency_level,
  starts_on,
  ends_on,
  application_url,
  source_url,
  metadata,
  is_active,
  updated_at,
  source_id,
  portal_id,
  event_id,
  organization:organizations!inner(
    id,
    name,
    slug,
    org_type,
    website,
    logo_url,
    description,
    city,
    neighborhood
  ),
  source:sources(
    id,
    name,
    slug,
    url
  ),
  event:events(
    id,
    title,
    start_date,
    start_time,
    end_time
  )
`;

function getDb(): AnySupabase {
  return createServiceClient() as unknown as AnySupabase;
}

async function getActivePortalBySlug(slug: string): Promise<PortalRow | null> {
  const db = getDb();
  const { data, error } = await db
    .from("portals")
    .select("id, slug, name")
    .eq("slug", slug)
    .in("status", ["active", "draft"])
    .maybeSingle();

  if (error) {
    throw new Error(`Failed to resolve portal: ${error.message}`);
  }

  return (data as PortalRow | null) ?? null;
}

async function getPortalAccessibleSourceIds(portalId: string): Promise<Set<number>> {
  const db = getDb();
  const { data, error } = await db
    .from("portal_source_access")
    .select("source_id")
    .eq("portal_id", portalId);

  if (error) {
    throw new Error(`Failed to resolve portal source access: ${error.message}`);
  }

  return new Set(
    ((data || []) as Array<{ source_id: number | null }>)
      .map((row) => row.source_id)
      .filter((value): value is number => typeof value === "number")
  );
}

function sortOpportunities<T extends VolunteerOpportunity>(opportunities: T[]): T[] {
  const urgencyWeight = (value: VolunteerOpportunity["urgency_level"]) => (value === "urgent" ? 0 : 1);
  const commitmentWeight = (value: VolunteerOpportunity["commitment_level"]) => {
    if (value === "lead_role") return 2;
    if (value === "ongoing") return 1;
    return 0;
  };
  const fitScore = (value: T) =>
    "fit_score" in value && typeof value.fit_score === "number" ? value.fit_score : 0;

  return [...opportunities].sort((a, b) => {
    const fitDiff = fitScore(b) - fitScore(a);
    if (fitDiff !== 0) return fitDiff;

    const urgencyDiff = urgencyWeight(a.urgency_level) - urgencyWeight(b.urgency_level);
    if (urgencyDiff !== 0) return urgencyDiff;

    const commitmentDiff = commitmentWeight(b.commitment_level) - commitmentWeight(a.commitment_level);
    if (commitmentDiff !== 0) return commitmentDiff;

    const updatedDiff = new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime();
    if (updatedDiff !== 0) return updatedDiff;

    return a.title.localeCompare(b.title);
  });
}

export function formatCommitmentLevel(value: VolunteerOpportunity["commitment_level"]): string {
  if (value === "lead_role") return "Lead";
  if (value === "ongoing") return "Ongoing";
  return "Drop-in";
}

export function formatOnboardingLevel(value: VolunteerOpportunity["onboarding_level"]): string | null {
  if (value === "screening_required") return "Screening required";
  if (value === "training_required") return "Training required";
  if (value === "light") return "Light onboarding";
  if (value === "none") return "No onboarding";
  return null;
}

export function formatTimeHorizon(value: VolunteerOpportunity["time_horizon"]): string | null {
  if (value === "multi_month") return "Multi-month";
  if (value === "multi_week") return "Multi-week";
  if (value === "one_day") return "One day";
  if (value === "ongoing") return "Open-ended";
  return null;
}

export function formatVolunteerCause(value: string): string {
  return value
    .split("_")
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(" ");
}

function getVolunteerOpportunityCause(opportunity: VolunteerOpportunity): string | null {
  return typeof opportunity.metadata?.cause === "string" ? opportunity.metadata.cause : null;
}

export function summarizeVolunteerOpportunities(
  opportunities: VolunteerOpportunity[],
): VolunteerOpportunitySummary {
  const byCommitmentLevel: VolunteerOpportunitySummary["by_commitment_level"] = {
    drop_in: 0,
    ongoing: 0,
    lead_role: 0,
  };
  const causeCounts = new Map<string, number>();

  for (const opportunity of opportunities) {
    byCommitmentLevel[opportunity.commitment_level] += 1;
    const cause = getVolunteerOpportunityCause(opportunity);
    if (cause) {
      causeCounts.set(cause, (causeCounts.get(cause) ?? 0) + 1);
    }
  }

  return {
    total: opportunities.length,
    by_commitment_level: byCommitmentLevel,
    by_cause: Array.from(causeCounts.entries())
      .map(([cause, count]) => ({ cause, count }))
      .sort((a, b) => b.count - a.count || a.cause.localeCompare(b.cause)),
  };
}

function matchesOpportunityFilters(
  opportunity: VolunteerOpportunity,
  filters: VolunteerOpportunityFilters,
): boolean {
  if (filters.cause) {
    const cause = getVolunteerOpportunityCause(opportunity);
    if (cause !== filters.cause) return false;
  }

  if (filters.query) {
    const haystack = [
      opportunity.title,
      opportunity.summary || "",
      opportunity.description || "",
      opportunity.organization.name,
      opportunity.organization.slug,
      opportunity.location_summary || "",
      opportunity.schedule_summary || "",
    ]
      .join(" ")
      .toLowerCase();
    if (!haystack.includes(filters.query.toLowerCase())) return false;
  }

  if (typeof filters.remoteAllowed === "boolean" && opportunity.remote_allowed !== filters.remoteAllowed) {
    return false;
  }

  if (filters.timeHorizon && opportunity.time_horizon !== filters.timeHorizon) {
    return false;
  }

  if (filters.onboardingLevel && opportunity.onboarding_level !== filters.onboardingLevel) {
    return false;
  }

  return true;
}

export async function getVolunteerProfileForUser(userId: string): Promise<VolunteerProfile | null> {
  const db = getDb();
  const { data, error } = await db
    .from("user_volunteer_profile")
    .select("*")
    .eq("user_id", userId)
    .maybeSingle();

  if (error) {
    throw new Error(`Failed to fetch volunteer profile: ${error.message}`);
  }

  return (data as VolunteerProfile | null) ?? null;
}

function rankVolunteerOpportunity(
  opportunity: VolunteerOpportunity,
  profile: VolunteerProfile | null | undefined,
): RankedVolunteerOpportunity {
  const reasons: string[] = [];
  let score = 0;

  if (!profile) {
    return {
      ...opportunity,
      fit_score: score,
      fit_reasons: reasons,
    };
  }

  const cause = getVolunteerOpportunityCause(opportunity);
  if (cause && profile.causes.includes(cause)) {
    score += 4;
    reasons.push("Cause match");
  }

  const skillOverlap = opportunity.skills_required.filter((skill) => profile.skills.includes(skill));
  if (skillOverlap.length > 0) {
    score += 3;
    reasons.push("Skill match");
  }

  if (profile.commitment_preference && profile.commitment_preference !== "mixed") {
    if (profile.commitment_preference === opportunity.commitment_level) {
      score += 2;
      reasons.push("Matches your commitment level");
    } else if (profile.commitment_preference === "ongoing" && opportunity.commitment_level === "lead_role") {
      score += 1;
      reasons.push("Near your commitment preference");
    }
  }

  if (profile.languages.some((language) => opportunity.language_support.includes(language))) {
    score += 2;
    reasons.push("Language match");
  }

  if (opportunity.remote_allowed && profile.mobility_constraints) {
    score += 1;
    reasons.push("Remote-friendly");
  }

  return {
    ...opportunity,
    fit_score: score,
    fit_reasons: reasons,
  };
}

export async function getVolunteerOpportunitiesForPortal(
  options: VolunteerOpportunityFilters & { profile?: VolunteerProfile | null },
): Promise<{
  portal: PortalRow;
  opportunities: RankedVolunteerOpportunity[];
  summary: VolunteerOpportunitySummary;
} | null> {
  const portal = await getActivePortalBySlug(options.portalSlug);
  if (!portal) return null;

  const db = getDb();
  const accessibleSourceIds = await getPortalAccessibleSourceIds(portal.id);

  let query = db
    .from("volunteer_opportunities")
    .select(VOLUNTEER_OPPORTUNITY_SELECT)
    .eq("is_active", true)
    .or(`portal_id.is.null,portal_id.eq.${portal.id}`);

  if (options.commitmentLevel) {
    query = query.eq("commitment_level", options.commitmentLevel);
  }

  if (options.organizationSlug) {
    query = query.eq("organization.slug", options.organizationSlug);
  }

  // Structured opportunity inventories are small enough that fetching a wider pool keeps
  // summary counts truthful even when a surface only renders a short ranked slice.
  const fetchLimit = Math.max(options.limit ?? 24, 200);
  const { data, error } = await query.limit(fetchLimit);
  if (error) {
    throw new Error(`Failed to fetch volunteer opportunities: ${error.message}`);
  }

  const filtered = ((data || []) as unknown as VolunteerOpportunity[]).filter((opportunity) => {
    if (opportunity.source_id === null) return true;
    if (accessibleSourceIds.size === 0) return true;
    return accessibleSourceIds.has(opportunity.source_id);
  }).filter((opportunity) => matchesOpportunityFilters(opportunity, options));

  const ranked = filtered.map((opportunity) => rankVolunteerOpportunity(opportunity, options.profile));

  return {
    portal,
    opportunities: sortOpportunities(ranked).slice(0, options.limit ?? 24),
    summary: summarizeVolunteerOpportunities(filtered),
  };
}

export async function getVolunteerOpportunityBySlug(options: {
  portalSlug: string;
  slug: string;
  profile?: VolunteerProfile | null;
}): Promise<{ portal: PortalRow; opportunity: RankedVolunteerOpportunity } | null> {
  const portal = await getActivePortalBySlug(options.portalSlug);
  if (!portal) return null;

  const db = getDb();
  const accessibleSourceIds = await getPortalAccessibleSourceIds(portal.id);
  const { data, error } = await db
    .from("volunteer_opportunities")
    .select(VOLUNTEER_OPPORTUNITY_SELECT)
    .eq("slug", options.slug)
    .eq("is_active", true)
    .or(`portal_id.is.null,portal_id.eq.${portal.id}`)
    .maybeSingle();

  if (error) {
    throw new Error(`Failed to fetch volunteer opportunity: ${error.message}`);
  }

  const opportunity = (data || null) as unknown as VolunteerOpportunity | null;
  if (!opportunity) return null;

  if (
    opportunity.source_id !== null
    && accessibleSourceIds.size > 0
    && !accessibleSourceIds.has(opportunity.source_id)
  ) {
    return null;
  }

  return { portal, opportunity: rankVolunteerOpportunity(opportunity, options.profile) };
}

export async function getVolunteerOpportunitiesForOrganization(options: {
  organizationSlug: string;
  limit?: number;
}): Promise<VolunteerOpportunity[]> {
  const db = getDb();
  const { data, error } = await db
    .from("volunteer_opportunities")
    .select(VOLUNTEER_OPPORTUNITY_SELECT)
    .eq("is_active", true)
    .eq("organization.slug", options.organizationSlug)
    .limit(options.limit ?? 12);

  if (error) {
    throw new Error(`Failed to fetch organization volunteer opportunities: ${error.message}`);
  }

  return sortOpportunities((data || []) as unknown as VolunteerOpportunity[]).slice(0, options.limit ?? 12);
}

File: crawlers/db/volunteer_opportunities.py
"""
Shared writes for volunteer_opportunities.

This gives HelpATL's structured opportunity family the same shared persistence
contract as other typed entity lanes instead of keeping it trapped inside
source-specific SQL snippets.
"""

import logging
import re
from typing import Optional

from db.client import (
    _log_write_skip,
    get_client,
    retry_on_network_error,
    writes_enabled,
)

logger = logging.getLogger(__name__)

_VOLUNTEER_OPPORTUNITY_COLUMNS = {
    "slug",
    "organization_id",
    "source_id",
    "portal_id",
    "event_id",
    "title",
    "summary",
    "description",
    "commitment_level",
    "time_horizon",
    "onboarding_level",
    "schedule_summary",
    "location_summary",
    "skills_required",
    "language_support",
    "physical_demand",
    "min_age",
    "family_friendly",
    "group_friendly",
    "remote_allowed",
    "accessibility_notes",
    "background_check_required",
    "training_required",
    "capacity_total",
    "capacity_remaining",
    "urgency_level",
    "starts_on",
    "ends_on",
    "application_url",
    "source_url",
    "metadata",
    "is_active",
}

_SLUG_RE = re.compile(r"[^\w\s-]")


def _slugify(text: str) -> str:
    slug = _SLUG_RE.sub("", text.lower())
    slug = re.sub(r"[\s_]+", "-", slug)
    slug = re.sub(r"-+", "-", slug).strip("-")
    return slug


@retry_on_network_error(max_retries=4, base_delay=0.5)
def _resolve_organization_id_by_slug(client, organization_slug: str):
    return (
        client.table("organizations")
        .select("id")
        .eq("slug", organization_slug)
        .limit(1)
        .execute()
    )


@retry_on_network_error(max_retries=4, base_delay=0.5)
def _upsert_volunteer_opportunity_record(client, row: dict):
    return client.table("volunteer_opportunities").upsert(
        row,
        on_conflict="slug",
    ).execute()


@retry_on_network_error(max_retries=4, base_delay=0.5)
def _select_active_source_opportunities(client, source_id: int):
    return (
        client.table("volunteer_opportunities")
        .select("id,slug")
        .eq("source_id", source_id)
        .eq("is_active", True)
        .execute()
    )


@retry_on_network_error(max_retries=4, base_delay=0.5)
def _deactivate_source_opportunities(client, stale_ids: list[str]):
    return (
        client.table("volunteer_opportunities")
        .update({"is_active": False})
        .in_("id", stale_ids)
        .execute()
    )


def upsert_volunteer_opportunity(opportunity_data: dict) -> Optional[str]:
    """Insert or update a structured volunteer opportunity matched by slug."""
    slug = (opportunity_data.get("slug") or "").strip()
    title = (opportunity_data.get("title") or "").strip()
    application_url = (opportunity_data.get("application_url") or "").strip()
    organization_id = opportunity_data.get("organization_id")
    organization_slug = (opportunity_data.get("organization_slug") or "").strip()

    if not slug:
        if title:
            slug = _slugify(title)[:120]
        else:
            logger.warning("upsert_volunteer_opportunity: missing slug and title")
            return None
    if not title:
        logger.warning("upsert_volunteer_opportunity: missing title for slug=%s", slug)
        return None
    if not application_url:
        logger.warning(
            "upsert_volunteer_opportunity: missing application_url for slug=%s",
            slug,
        )
        return None

    client = get_client()
    if not organization_id:
        if not organization_slug:
            logger.warning(
                "upsert_volunteer_opportunity: missing organization reference for slug=%s",
                slug,
            )
            return None
        try:
            organization_rows = (
                _resolve_organization_id_by_slug(client, organization_slug).data or []
            )
        except Exception:
            logger.exception(
                "Failed to resolve organization slug=%s for volunteer opportunity slug=%s",
                organization_slug,
                slug,
            )
            return None
        if not organization_rows:
            logger.warning(
                "upsert_volunteer_opportunity: unknown organization_slug=%s for slug=%s",
                organization_slug,
                slug,
            )
            return None
        organization_id = organization_rows[0]["id"]

    row = {
        "slug": slug,
        "organization_id": organization_id,
        **{
            key: value
            for key, value in opportunity_data.items()
            if key in _VOLUNTEER_OPPORTUNITY_COLUMNS
            and key not in {"slug", "organization_id"}
        },
    }
    row["title"] = title
    row["application_url"] = application_url
    row["commitment_level"] = row.get("commitment_level") or "ongoing"
    row["metadata"] = row.get("metadata") or {}
    if "is_active" not in row:
        row["is_active"] = True

    if not writes_enabled():
        _log_write_skip(
            f"upsert volunteer_opportunities slug={slug} organization_id={organization_id}"
        )
        return slug

    try:
        result = _upsert_volunteer_opportunity_record(client, row)
        if result.data:
            record_id = result.data[0].get("id")
            if isinstance(record_id, str) and record_id:
                return record_id
            return slug
    except Exception:
        logger.exception(
            "Failed to upsert volunteer opportunity slug=%s organization_id=%s",
            slug,
            organization_id,
        )
        return None

    return None


def deactivate_stale_volunteer_opportunities(
    source_id: int,
    active_slugs: set[str],
) -> int:
    """Deactivate active source-owned opportunities missing from the current run."""
    client = get_client()
    rows = (_select_active_source_opportunities(client, source_id).data or [])
    stale_ids = [row["id"] for row in rows if row["slug"] not in active_slugs]
    if not stale_ids or not writes_enabled():
        return len(stale_ids)

    try:
        _deactivate_source_opportunities(client, stale_ids)
        return len(stale_ids)
    except Exception:
        logger.exception(
            "Failed to deactivate stale volunteer opportunities for source_id=%s",
            source_id,
        )
        return 0
